gapii/cc/vulkan_external_memory.cpp

```diff
@@ -0,0 +1,497 @@
+      ExternalMemoryCommandBuffer& cmdBuf = submits[i].commandBuffers[j];
+      cmdBuf.commandBuffer = pSubmits[i].mpCommandBuffers[j];
+      auto bufIt = spy->mExternalBufferBarriers.find(cmdBuf.commandBuffer);
+      if (bufIt != spy->mExternalBufferBarriers.end()) {
+        for (auto barrierIt = bufIt->second.begin();
+             barrierIt != bufIt->second.end(); ++barrierIt) {
+          cmdBuf.buffers.push_back(
+              ExternalBufferMemoryStaging(*barrierIt, stagingSize));
+        }
+      }
+    }
+  }
+}
+
+uint32_t ExternalMemoryStaging::CreateResources() {
+  VkCommandPoolCreateInfo commandPoolCreateInfo{
+      VkStructureType::VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO,  // sType
+      nullptr,                                                      // pNext
+      VkCommandPoolCreateFlagBits::
+          VK_COMMAND_POOL_CREATE_TRANSIENT_BIT,  // flags
+      queueFamilyIndex,                          // queueFamilyIndex
+  };
+  uint32_t res = fn->vkCreateCommandPool(device, &commandPoolCreateInfo,
+                                         nullptr, &stagingCommandPool);
+  if (res != VkResult::VK_SUCCESS) {
+    stagingCommandPool = 0;
+    GAPID_ERROR("Error creating command pool for external memory observations");
+    return res;
+  }
+
+  VkFenceCreateInfo fenceCreateInfo{
+      VkStructureType::VK_STRUCTURE_TYPE_FENCE_CREATE_INFO,  // sType
+      nullptr,                                               // pNext
+      0,                                                     // flags
+  };
+  res = fn->vkCreateFence(device, &fenceCreateInfo, nullptr, &stagingFence);
+  if (res != VkResult::VK_SUCCESS) {
+    stagingFence = 0;
+    GAPID_ERROR("Error creating fence for external memory observations");
+    return res;
+  }
+
+  size_t commandBufferCount = 1;
+  for (auto submitIt = submits.begin(); submitIt != submits.end(); ++submitIt) {
+    for (auto cmdBufIt = submitIt->commandBuffers.begin();
+         cmdBufIt != submitIt->commandBuffers.end(); ++cmdBufIt) {
+      if (!cmdBufIt->empty()) {
+        ++commandBufferCount;
+      }
+    }
+  }
+  std::vector<VkCommandBuffer> commandBuffers(commandBufferCount);
+  VkCommandBufferAllocateInfo commandBufferAllocInfo{
+      VkStructureType::VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO,  // sType
+      nullptr,                                                          // pNext
+      stagingCommandPool,                                     // commandPool
+      VkCommandBufferLevel::VK_COMMAND_BUFFER_LEVEL_PRIMARY,  // level
+      (uint32_t)commandBuffers.size() + 1,  // commandBufferCount
+  };
+  res = fn->vkAllocateCommandBuffers(device, &commandBufferAllocInfo,
+                                     commandBuffers.data());
+  if (res != VkResult::VK_SUCCESS) {
+    GAPID_ERROR(
+        "Error allocating command buffer for external memory observations");
+    return res;
+  }
+  stagingCommandBuffer = commandBuffers.back();
```

Comment:
I think you may have to fix up the dispatch table here. 
see   `set_dispatch_from_parent((void*)command_buffer, (void*)device);`

That way other layers will play nice with while GAPID is enabled.

Reply:
I think that is a perfectly fine place